onnxruntime/core/providers/qnn/builder/qnn_node_group/reshape_gemm_fusion.cc

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/qnn/builder/qnn_node_group/reshape_gemm_fusion.h"
+
+#include <gsl/gsl>
+#include <algorithm>
+#include <cassert>
+#include <limits>
+#include <optional>
+#include <string>
+
+#include "core/graph/graph_utils.h"
+#include "core/framework/node_unit.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/providers/qnn/builder/qnn_utils.h"
+#include "core/providers/qnn/builder/op_builder_factory.h"
+#include "core/providers/qnn/builder/qnn_node_group/utils.h"
+#include "core/providers/qnn/builder/qnn_model_wrapper.h"
+#include "core/providers/shared/utils/utils.h"
+
+namespace onnxruntime {
+namespace qnn {
+
+namespace {
+
+const NodeUnit* GetReshapeNodeUnit(
+    const GraphViewer& graph_viewer, const std::unordered_map<const Node*, const NodeUnit*>& node_to_node_unit,
+    const std::unordered_map<const NodeUnit*, const IQnnNodeGroup*>& node_unit_to_qnn_node_group,
+    const Node& gemm_node) {
+  if (gemm_node.OpType() != "Gemm") {
+    return nullptr;
+  }
+
+  for (auto edge = gemm_node.InputEdgesBegin(); edge != gemm_node.InputEdgesEnd(); edge++) {
+    if (edge->GetDstArgIndex() == 0) {
+      const Node& reshape_node = edge->GetNode();
+      if (reshape_node.OpType() == "Reshape" && !graph_viewer.NodeProducesGraphOutput(reshape_node) &&
+          reshape_node.GetOutputEdgesCount() == 1) {
+        const auto it = node_to_node_unit.find(&reshape_node);
+        if (it != node_to_node_unit.end()) {
+          const NodeUnit* reshape_node_unit = it->second;
+          if (reshape_node_unit && node_unit_to_qnn_node_group.count(reshape_node_unit) == 0 &&
+              reshape_node_unit->UnitType() == NodeUnit::Type::SingleNode) {
+            return reshape_node_unit;
+          }
+        }
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+// Reshape from [x0, x1, ..., xn, k] to [x0 * x1 * ... * xn, k].
+bool CheckShape(const Node& reshape_node) {
+  auto input_shape_proto = reshape_node.InputDefs()[0]->Shape();
+  auto output_shape_proto = reshape_node.OutputDefs()[0]->Shape();
+  if (!input_shape_proto || !output_shape_proto) {
+    return false;
+  }
+
+  auto input_shape = onnxruntime::utils::GetTensorShapeFromTensorShapeProto(*input_shape_proto);
+  auto output_shape = onnxruntime::utils::GetTensorShapeFromTensorShapeProto(*output_shape_proto);
```

Comment:
Hi @centwang, Apologies for not getting to this PR sooner. We recently merged a commit that made QNN EP a shared library by default. This PR now has some conflicts with main. Let me know if I can help you with getting this PR sync'd with main.

Reply:
Sure, please go ahead.